modules/message-lane/src/outbound_lane.rs

```diff
@@ -0,0 +1,269 @@
+
+use bp_message_lane::{LaneId, MessageNonce, OutboundLaneData};
+
+/// Outbound lane storage.
+pub trait OutboundLaneStorage {
+	/// Message payload.
+	type Payload;
+
+	/// Lane id.
+	fn id(&self) -> LaneId;
+	/// Get lane data from the storage.
+	fn data(&self) -> OutboundLaneData;
+	/// Update lane data in the storage.
+	fn set_data(&mut self, data: OutboundLaneData);
+	/// Returns saved outbound message payload.
+	#[cfg(test)]
+	fn message(&self, nonce: &MessageNonce) -> Option<Self::Payload>;
+	/// Save outbound message in the storage.
+	fn save_message(&mut self, nonce: MessageNonce, payload: Self::Payload);
+	/// Remove outbound message from the storage.
+	fn remove_message(&mut self, nonce: &MessageNonce);
+}
+
+/// Outbound messages lane.
+pub struct OutboundLane<S> {
+	storage: S,
+}
+
+impl<S: OutboundLaneStorage> OutboundLane<S> {
+	/// Create new inbound lane backed by given storage.
+	pub fn new(storage: S) -> Self {
+		OutboundLane { storage }
+	}
+
+	/// Send message over lane.
+	///
+	/// Returns new message nonce.
+	pub fn send_message(&mut self, payload: S::Payload) -> MessageNonce {
+		let mut data = self.storage.data();
+		let nonce = data.latest_generated_nonce + 1;
+		data.latest_generated_nonce = nonce;
+
+		self.storage.save_message(nonce, payload);
+		self.storage.set_data(data);
+
+		nonce
+	}
+
+	/// Confirm message receival.
+	///
+	/// Returns `None` if confirmation is wrong/duplicate.
+	/// Returns `Some` with inclusive ranges of message nonces that have been received.
+	pub fn confirm_receival(&mut self, latest_received_nonce: MessageNonce) -> Option<(MessageNonce, MessageNonce)> {
+		let mut data = self.storage.data();
+		if latest_received_nonce <= data.latest_received_nonce || latest_received_nonce > data.latest_generated_nonce {
+			return None;
+		}
+
+		let prev_latest_received_nonce = data.latest_received_nonce;
+		data.latest_received_nonce = latest_received_nonce;
+		self.storage.set_data(data);
+
+		Some((prev_latest_received_nonce + 1, latest_received_nonce))
+	}
+
+	/// Confirm message processing.
+	///
+	/// Returns `None` if confirmation is wrong/duplicate.
+	/// Returns `Some` with inclusive ranges of message nonces that have been processed.
+	pub fn confirm_processing(&mut self, latest_processed_nonce: MessageNonce) -> Option<(MessageNonce, MessageNonce)> {
+		let mut data = self.storage.data();
+		// wait for recieval confirmation first
+		if latest_processed_nonce <= data.latest_processed_nonce || latest_processed_nonce > data.latest_received_nonce
+		{
+			return None;
+		}
+
+		let prev_latest_processed_nonce = data.latest_processed_nonce;
+		data.latest_processed_nonce = latest_processed_nonce;
+		self.storage.set_data(data);
+
+		Some((prev_latest_processed_nonce + 1, latest_processed_nonce))
+	}
+
+	/// Prune at most `max_messages_to_prune` already received messages.
+	pub fn prune_messages(&mut self, max_messages_to_prune: MessageNonce) -> MessageNonce {
```

Comment:
This doesn't return the message nonce, it returns the number of pruned messages. It's subtle, but it tripped me up when reading the pruning test.

Reply:
Updated doc